### PR TITLE
Replace PID with Task Pointer in Rusty

### PR DESCRIPTION
**Pull request by likewhatevs**
~I wasnt't able to reopen https://github.com/sched-ext/scx/pull/672 because I force pushed (just learned about that limit).~

~This is a continuation of that PR.~

~I closed that PR and opened https://github.com/sched-ext/scx/pull/710 because, while the first PR's approach passed stress tests and seemed to work, it kind of didn't make sense that I could deref the casted pointer to task_struct.~

~I updated that initial PR to explicltly deref that casted pointer to confirm this works/the verifier allows it and it seems to work so, this works?~

see latest comment

### CODE REVIEW

File: scheds/rust/scx_rusty/src/bpf/main.bpf.c [hunk 1]
```diff
@@ -417,7 +417,9 @@ static int dom_xfer_task(u64 task_id, u32 new_dom_id, u64 now)
 	if (!get_task_from_id(task_id, p)) {
 		scx_bpf_error("Failed to lookup task_struct");
 		return 0;
-	}
+	} 
+	
+	s32 a = p->pid;
```
Comment (htejun) [inline scheds/rust/scx_rusty/src/bpf/main.bpf.c:422]: Compiler will most likely eliminate this.

**Comment by vax-r** [comment]
Wouldn't this design throw away the origin safety provided by `bpf_task_from_pid()` ? It was able to perform reference count on the `task_struct` .
Using a map to store is awesome but it seems that we do not have mechanism to cleanup the unused <k, v> pair in the map, as the time going, the map size can grow really big, even it's acceptable, we would still like to avoid redundant memory usage , maybe having that implemented will be nice?

**Comment by vax-r** [comment]
Btw I want to ask what's the actual cause of `Failed to lookup task %d` ?

**Comment by likewhatevs** [comment]
I *think* this is the best we (or well, at least I) can do atm. 

So like, I tried a few ways of tracking the association between the initial pid of a task struct and the pointer of the task struct. 

My most recent attempts errored at null-checking the task context read using that association:
```
; if (!taskc) goto free_task; @ main.bpf.c:456
372: (15) if r3 == 0x0 goto pc+27
R3 !read_ok
processed 11065 insns (limit 1000000) max_states_per_insn 8 total_states 740 peak_states 298 mark_read 18
-- END PROG LOAD LOG --
```

Unless I'm reading something wrong or misunderstand something, I think this is the best we can do ATM (i.e. until it's possible to null check the retrieved variable).

I think this might be alright though because like, tests *do* pass and logs on those tests do not have stalls (which, I think, is the main reason to be concerned we are tracking everything).

**Comment by vax-r** [comment]
> I think this might be alright though because like, tests _do_ pass and logs on those tests do not have stalls (which, I think, is the main reason to be concerned we are tracking everything).

It seems to me that what the commit does is only refactor the origin code , rather than actually changing any logic or mechanism when dealing with `pid` and `task_struct`, how can it pass this time ? It seems confusing to me.

**Comment by hodgesds** [comment]
> This is correct. What I am wondering is if the current behavior (without that exception being thrown) is problematic. I would think I would see a stall if it were if I understand how things work correctly now, but I'm not super certain of that.

I think the idea is that if correct scheduling decisions can't be made by the the scheduler it's better to exit (the scheduler) rather than making bad scheduling decisions.  There's still some bugs in rusty, so in some ways it maybe is better to fail fast in more places. In this case of looking up the task context I think it's probably ok to bail the scheduler, because if the task context is lost/incorrect then making correct scheduling decisions is probably not handled correctly elsewhere.

**Comment by hodgesds** [comment]
> This is the part where like, IDK. So like, if we lost the task, it wouldn't be scheduled, and we would see a stall?

I think the problem with `stress-ng` tests is that with the `--fork` tests the tasks come and go really fast, I'm not sure if there's an easy way to trigger the behavior (maybe compiling the kernel like in #684) and see what happens with this change. Maybe do a kernel build with something crazy like ````make -j $(nproc * 2)```.

**Comment by likewhatevs** [comment]
> I think the problem with stress-ng tests is that with the --fork tests the tasks come and go really fast, I'm not sure if there's an easy way to trigger the behavior (maybe compiling the kernel like in https://github.com/sched-ext/scx/issues/684) and see what happens with this change. Maybe do a kernel build with something crazy like ````make -j $(nproc * 2)```.

gonna run it on laptop for a day and do the terraria/compile test/see if I can find bad behavior, good idea, thx!

**Comment by likewhatevs** [comment]
Alright so I think I finally got this (and learned a ton along the way).

This PR now has Rusty use Task pointer casted to u64 to track tasks instead of pid.

A couple things I learned:

In bpf, you can't cast "trusted pointers". I don't fully get it, but t_to_tptr makes the cast work w/ the verifier.

Making functions static makes some problems go away, but can also super quickly drive things over the instruction limit.

*More interestingly*, I tried only grabbing pid via probe_read (rather than a pointer to task_struct), and to then grab task_struct via `bpf_task_from_pid` (because maybe the bpf locking could be useful/I was curious). 

What I found is that is *super* racy (i.e. adding a counter and a while loop made stress test last longer while before failing with a missing pid error), so I kind of think we need to not use `bpf_task_from_pid` anywhere.

**Review by hodgesds — APPROVED** [review]
I think this will work!

File: scheds/rust/scx_rusty/src/bpf/main.bpf.c [hunk 2]
```diff
@@ -397,31 +414,43 @@ static u64 dom_min_vruntime(struct dom_ctx *domc)
...
+    struct task_struct *p;
+	int err_task;
+	struct task_struct *task;
+
+	task = (struct task_struct *)tptr;
+
+	err_task = bpf_probe_read_kernel(&p, sizeof(struct task_struct *), &task);
+
+	if (err_task) scx_bpf_error("Failed to retrieve task_struct for tptr %llu", tptr);
+	if(p) return p;
```
Comment (hodgesds) [inline scheds/rust/scx_rusty/src/bpf/main.bpf.c:427]: nit: consistent spacing

File: scheds/rust/scx_rusty/src/bpf/main.bpf.c [hunk 3]
```diff
@@ -397,31 +414,43 @@ static u64 dom_min_vruntime(struct dom_ctx *domc)
 	return READ_ONCE(domc->min_vruntime);
 }
 
-int dom_xfer_task(pid_t pid, u32 new_dom_id, u64 now)
+static struct task_struct *tptr_to_task(u64 tptr){
```
Comment (hodgesds) [inline scheds/rust/scx_rusty/src/bpf/main.bpf.c:417]: nit:
```
static struct task_struct *tptr_to_task(u64 tptr)
{
        struct task_struct *p, *task;
	int err_task;
```

File: scheds/rust/scx_rusty/src/bpf/main.bpf.c [hunk 4]
```diff
@@ -788,9 +819,10 @@ static bool task_set_domain(struct task_ctx *taskc, struct task_struct *p,
 	if (bpf_cpumask_intersects((const struct cpumask *)d_cpumask,
 				   p->cpus_ptr)) {
 		u64 now = bpf_ktime_get_ns();
-
-		if (!init_dsq_vtime)
-			dom_xfer_task(p->pid, new_dom_id, now);
+		
+		if (!init_dsq_vtime){
```
Comment (hodgesds) [inline scheds/rust/scx_rusty/src/bpf/main.bpf.c:823]: nit:
```
if (!init_dsq_vtime) {
```

**Pull request merged**